lib/THC/THCTensorCopy.cu

```diff
@@ -28,62 +28,122 @@ THCudaTensor_copy(THCState* state, THCudaTensor* dst, THCudaTensor* src) {
+    cudaEvent_t dstReady;
+    THCudaCheck(cudaSetDevice(dstDev));
+    THCudaCheck(cudaEventCreateWithFlags(&dstReady, cudaEventDisableTiming));
+    THCudaCheck(cudaEventRecord(dstReady, NULL));
+
+    THCudaCheck(cudaSetDevice(srcDev));
+    THCudaCheck(cudaStreamWaitEvent(NULL, dstReady, 0));
+    THCudaCheck(cudaEventDestroy(dstReady));
+  } else if (srcDev != oldDev) {
+    THCudaCheck(cudaSetDevice(srcDev));
   }
 
+  // We are now on srcDev
   if (memcpyEligible) {
+    // Perform the copy
     THCudaCheck(cudaMemcpyAsync(THCudaTensor_data(state, dst),
                                 THCudaTensor_data(state, src),
                                 totalElements * sizeof(float),
                                 cudaMemcpyDeviceToDevice,
-                                THCState_getCurrentStream(state)));
+                                copyStream));
   } else {
+    // Non-contiguous copy
+
+    // We avoid creating temporary memory copies if possible.
+    // If both src and dst are on the same device, or if they are on
+    // different devices and p2p access is enabled, perform the copy
+    // by a pointwise copy kernel.
+    // Otherwise, we'll have to make contiguous (which will in fact
+    // invoke copy() again), and then perform the copy.
+    // FIXME: might want to consider only running the pointwise kernel
+    // if both src and dst innermost dimensions are contiguous. If
+    // they are not, then taking the hit of the memory allocation/free
+    // might be worth it to avoid non-coalesced reads or writes.
+    if ((srcDev == oldDev) ||
+        THCState_peerToPeerAccessEnabled(state)) {
+      // Make sure we have the current stream set in THCState, since
+      // pointwise uses that
+      if (srcDev != oldDev) {
+        THCState_setStream(state, srcDev, copyStreamIndex);
+      }
+
       bool succ =
         THCudaTensor_pointwiseApply2(state, dst, src, CopyOp<float>());
       THArgCheck(succ, 2, CUTORCH_DIM_WARNING);
-  }
 
-  if (srcDev != dstDev) {
-    // synchronize remote device after copy
-    cudaEvent_t doneCopying;
-    THCudaCheck(cudaEventCreate(&doneCopying));
-    THCudaCheck(cudaEventRecord(
-                  doneCopying,
-                  THCState_getDeviceStream(state, copyDev, THCState_getCurrentStreamIndex(state))));
-    THCudaCheck(cudaSetDevice(remoteDev));
-    THCudaCheck(cudaStreamWaitEvent(
-                  THCState_getDeviceStream(state, remoteDev, THCState_getCurrentStreamIndex(state)),
-                  doneCopying, 0));
-    THCudaCheck(cudaEventDestroy(doneCopying));
+      // Restore prior THCState stream
+      if (srcDev != oldDev) {
+        THCState_setStream(state, oldDev, copyStreamIndex);
+      }
+    } else {
+      // GPUs can't access each other directly; fall back to
+      // newContiguous and memcpy
+      THCudaTensor* srcContig = THCudaTensor_newContiguous(state, src);
+      THCudaTensor* dstContig = THCudaTensor_newContiguous(state, dst);
```

Comment:
the copy is unavoidable, without p2p you can't copy using a kernel if dst is non-contig, you have to use cudaMemcpyAsync which will route the copy via the host. if dst is contig then it just addrefs it.

in theory one could see if both are non-contiguous in the same manner and if there are no holes and one can do the copy, but that might be an excessive check